consensus/obcpbft/pbft-core.go

```diff
@@ -249,67 +256,60 @@ func newPbftCore(id uint64, config *viper.Viper, consumer innerStack) *pbftCore
 	instance.outstandingReqs = make(map[string]*Request)
 	instance.missingReqs = make(map[string]bool)
 
-	instance.stopTimer()
 	instance.restoreState()
 
-	go instance.main()
-
 	return instance
 }
 
 // close tears down resources opened by newPbftCore
 func (instance *pbftCore) close() {
-	select { // Prevent closing multiple times
-	case <-instance.closed:
-	default:
-		close(instance.closed)
-	}
-	instance.stopTimer()
+	instance.manager.halt()
+	instance.newViewTimer.halt()
 }
 
 // allow the view-change protocol to kick-off when the timer expires
-func (instance *pbftCore) main() {
+func (instance *pbftCore) processEvent(e event) event {
 
-	for {
-		logger.Debug("Replica %d service thread looping", instance.id)
+	logger.Debug("Replica %d processing event", instance.id)
 
-		select {
-		case <-instance.closed:
-			close(instance.idleChan)
-			return
-		case <-instance.newViewTimer.C:
-			logger.Info("Replica %d view change timer expired, sending view change", instance.id)
-			instance.sendViewChange()
-		case msg := <-instance.incomingChan:
-			logger.Debug("Replica %d received incoming message from %v", instance.id, msg.sender)
-			instance.recvMsg(msg.msg, msg.sender)
-		case update := <-instance.stateUpdatingChan:
-			instance.skipInProgress = true
-			instance.lastExec = update.seqNo
-			instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
-		case update := <-instance.stateUpdatedChan:
-			seqNo := update.seqNo
-			logger.Info("Replica %d application caught up via state transfer, lastExec now %d", instance.id, seqNo)
-			// XXX create checkpoint
-			instance.lastExec = seqNo
-			instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
-			instance.skipInProgress = false
-			instance.executeOutstanding()
-		case <-instance.execCompleteChan:
-			instance.execDoneSync()
-		case work := <-instance.injectChan:
-			work() // Used to allow the caller to steal use of the main thread
-		case instance.idleChan <- struct{}{}:
-			// Used to detect when this thread is idle for testing
-		}
+	switch e.eventType() {
+	case viewChangeTimerEventID:
+		logger.Info("Replica %d view change timer expired, sending view change", instance.id)
+		instance.timerActive = false
+		instance.sendViewChange()
+	case messageEventID:
+		msg := e.(messageEvent)
+		logger.Debug("Replica %d received incoming message from %v", instance.id, msg.sender)
+		instance.recvMsg(msg.msg, msg.sender)
+	case stateUpdatingEventID:
+		update := e.(stateUpdatingEvent)
+		instance.skipInProgress = true
+		instance.lastExec = update.seqNo
+		instance.moveWatermarks(instance.lastExec) // The watermark movement handles moving this to a checkpoint boundary
+	case stateUpdatedEventID:
+		update := e.(stateUpdatedEvent)
```

Comment:
As mentioned before, this is stopgap, the original implementation was type inference, and I plan to move back to it.